src/ops/euclid.rs

```diff
@@ -0,0 +1,327 @@
+use core::ops::{Div, Rem};
+pub trait DivEuclid: Sized + Div<Self, Output = Self> {
+    /// Calculates Euclidean division, the matching method for `rem_euclid`.
+    ///
+    /// This computes the integer `n` such that
+    /// `self = n * v + self.rem_euclid(v)`.
+    /// In other words, the result is `self / v` rounded to the integer `n`
+    /// such that `self >= n * v`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use num_traits::DivEuclid;
+    ///
+    /// let a: i32 = 7;
+    /// let b: i32 = 4;
+    /// assert_eq!(DivEuclid::div_euclid(a,b), 1); // 7 > 4 * 1
+    /// assert_eq!(DivEuclid::div_euclid(-a,b), -2); // -7 >= 4 * -2
+    /// assert_eq!(DivEuclid::div_euclid(a,-b), -1); // 7 >= -4 * -1
+    /// assert_eq!(DivEuclid::div_euclid(-a,-b), 2); // -7 >= -4 * 2
+    /// ```
+    fn div_euclid(self, v: Self) -> Self;
+}
+pub trait RemEuclid: Sized + Rem<Self, Output = Self> {
+    /// Calculates the least nonnegative remainder of `self (mod v)`.
+    ///
+    /// In particular, the return value `r` satisfies `0.0 <= r < v.abs()` in
+    /// most cases. However, due to a floating point round-off error it can
+    /// result in `r == v.abs()`, violating the mathematical definition, if
+    /// `self` is much smaller than `v.abs()` in magnitude and `self < 0.0`.
+    /// This result is not an element of the function's codomain, but it is the
+    /// closest floating point number in the real numbers and thus fulfills the
+    /// property `self == self.div_euclid(v) * v + self.rem_euclid(v)`
+    /// approximatively.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use num_traits::RemEuclid;
+    ///
+    /// let a: i32 = 7;
+    /// let b: i32 = 4;
+    /// assert_eq!(RemEuclid::rem_euclid(a,b), 3);
+    /// assert_eq!(RemEuclid::rem_euclid(-a,b), 1);
+    /// assert_eq!(RemEuclid::rem_euclid(a,-b), 3);
+    /// assert_eq!(RemEuclid::rem_euclid(-a,-b), 1);
+    /// ```
+    fn rem_euclid(self, v: Self) -> Self;
+}
+macro_rules! div_euclid_int_impl {
+    ($trait_name:ident for $($t:ty)*) => {$(
```

Comment:
The parameterized `$trait_name` isn't helping us here. I think this should hardcode the trait name and combine `div`/`rem`, so it can be simply invoked `euclid_int_impl(i8 i16 i32 i64 isize)`.